bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java

```diff
@@ -654,42 +624,28 @@ public void readEntryAndFenceLedger(final long ledgerId, byte[] masterKey,
         }
 
         final CompletionKey completionKey = completion;
-        if (completionObjects.putIfAbsent(completionKey, new ReadCompletion(this, readEntryOpLogger, cb,
-                ctx, ledgerId, entryId, scheduleTimeout(completionKey, readEntryTimeout))) != null) {
+        if (completionObjects.putIfAbsent(
```

Comment:
I wonder if we can save some allocation using computeIfAbsent...maybe not

Reply:
computeIfAbsent would still need a Function allocated, so you don't really save anything.

Reply:
@ivankelly 
you are right, we need to capture the context so there is no way of not allocating a new object